subsys/net/lib/coap/coap_server.c

```diff
@@ -0,0 +1,762 @@
+	if (ret < 0) {
+		LOG_ERR("Failed to send CoAP message (%d)", ret);
+		return ret;
+	}
+	__ASSERT_NO_MSG(ret == cpkt->offset);
+
+	return 0;
+}
+
+int coap_resource_send(const struct coap_resource *resource, const struct coap_packet *cpkt,
+		       const struct sockaddr *addr, socklen_t addr_len)
+{
+	/* Find owning service */
+	COAP_SERVICE_FOREACH(svc) {
+		if (COAP_SERVICE_HAS_RESOURCE(svc, resource)) {
+			return coap_service_send(svc, cpkt, addr, addr_len);
+		}
+	}
+
+	return -ENOENT;
+}
+
+int coap_resource_parse_observe(struct coap_resource *resource, const struct coap_packet *request,
+				const struct sockaddr *addr)
+{
+	const struct coap_service *service = NULL;
+	int ret;
+
+	if (!coap_packet_is_request(request)) {
+		return -EINVAL;
+	}
+
+	ret = coap_get_option_int(request, COAP_OPTION_OBSERVE);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Find owning service */
+	COAP_SERVICE_FOREACH(svc) {
+		if (COAP_SERVICE_HAS_RESOURCE(svc, resource)) {
+			service = svc;
+			break;
+		}
+	}
+
+	if (service == NULL) {
+		return -ENOENT;
+	}
+
+	(void)k_mutex_lock(&lock, K_FOREVER);
+
+	if (ret == 0) {
+		struct coap_observer *observer;
+
+		observer = coap_observer_next_unused(service->data->observers, MAX_OBSERVERS);
+		if (observer == NULL) {
+			ret = -ENOMEM;
+			goto unlock;
+		}
+
+		coap_observer_init(observer, request, addr);
+		coap_register_observer(resource, observer);
+	} else if (ret == 1) {
+		uint8_t token[COAP_TOKEN_MAX_LEN];
+		uint8_t tkl;
+
+		tkl = coap_header_get_token(request, token);
+		ret = coap_service_remove_observer(service, resource, addr, token, tkl);
+		if (ret < 0) {
+			LOG_WRN("Failed to remove observer (%d)", ret);
+		}
+	}
+
+unlock:
+	(void)k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+static int coap_resource_remove_observer(struct coap_resource *resource,
+					 const struct sockaddr *addr,
+					 const uint8_t *token, uint8_t token_len)
+{
+	const struct coap_service *service = NULL;
+	int ret;
+
+	/* Find owning service */
+	COAP_SERVICE_FOREACH(svc) {
+		if (COAP_SERVICE_HAS_RESOURCE(svc, resource)) {
+			service = svc;
+			break;
+		}
+	}
+
+	if (service == NULL) {
+		return -ENOENT;
+	}
+
+	(void)k_mutex_lock(&lock, K_FOREVER);
+	ret = coap_service_remove_observer(service, resource, addr, token, token_len);
+	(void)k_mutex_unlock(&lock);
+
+	if (ret == 1) {
+		/* An observer was found and removed */
+		return 0;
+	} else if (ret == 0) {
+		/* No matching observer found */
+		return -ENOENT;
+	}
+
+	/* An error occurred */
+	return ret;
+}
+
+int coap_resource_remove_observer_by_addr(struct coap_resource *resource,
+					  const struct sockaddr *addr)
+{
+	return coap_resource_remove_observer(resource, addr, NULL, 0);
+}
+
+int coap_resource_remove_observer_by_token(struct coap_resource *resource,
+					   const uint8_t *token, uint8_t token_len)
+{
+	return coap_resource_remove_observer(resource, NULL, token, token_len);
+}
+
+static void coap_server_thread(void *p1, void *p2, void *p3)
```

Comment:
Although this server lib is kind of doing similar things as what http server PR (#64465) is doing, it differs here. So we create a thread in library, and then it does the magic. In the http server lib, there is `http_server_init()` which loops through services that need to be created, and then opens the listening sockets. Then in `http_server_start()` it starts to wait client connections.
But it leaves the thread creation (which imho is kind of application related thing) to the application. Did you consider something like that here. Because this is a library, creating threads here is a bit problematic. I would leave as much of resource allocation (like threads and stacks) to the application side, as it is kind of messy configuration and mgmt wise.

@cfriedt you have been involved with the http server design from the beginning, what is you take on this?

Reply:
I did think about this, the main difference here, is that CoAP is UDP based and retransmissions are handled by the subsystem/application layer.
Having a dedicated thread for the application logic did simplify writing the server logic. But I do understand your concern, maybe @cfriedt can help indeed.

Reply:
As a side note, the CoAP client library also uses a separate thread in the subsystem.

Reply:
> Although this server lib is kind of doing similar things as what http server PR (#64465) is doing, it differs here.

It could be the case that #64465 (and the original GSoC code) still are not implementing the asynchronous server that was intended.

With an asynchronous server, there would not need to be any dedicated thread.

On the other hand, it is definitely simpler to use a thread and blocking I/O.

> `http_server_start()` it starts to wait client connections.

So that's where it went off course in a way. For an asynchronous server, we would have a kind of "tick" function for the server to call regularly (or on a per-event basis). Technically, a thread could be used even in that scenario, but all of would do is poll for events and then call the "tick" function.

I would suggest aiming for that kind of architecture in the long run, because technically, it is quite scalable and versatile. Of course, there will likelu be more statefulness. 

> @cfriedt you have been involved with the http server design from the beginning, what is you take on this?

I'm all for functional-first, so if the threaded server works, then that's ok. Optimization is a long process (sometimes too long).

Given that CoAP is (IIRC) a datagram-oriented protocol, I think it's significantly safer than HTTP in terms of resilience to DOS and timing attacks. Still have to take care for security and robustness of course.